File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Models/_helpers/rateable.py
from django.db import models
from .._helpers.default_timed import DefaultTimed
from .._helpers.fields import (default_user_creation, default_char_field, uuid_def_primary, 
                               EVAL_CHOICES, VOID)
class Rateable(DefaultTimed):
  
  class Meta:
    abstract = True

  id = models.UUIDField(**uuid_def_primary())
  integrity = models.TextField(max_length=2048)
  integrity_rating = default_char_field(
      choices=EVAL_CHOICES,
      default='average'
  )
  attendance = models.TextField(max_length=2048)
  attendance_rating = default_char_field(
      choices=EVAL_CHOICES,
      default='average'
  )
  customer_experience = models.TextField(max_length=2048)
  customer_experience_rating = default_char_field(
      choices=EVAL_CHOICES,
      default='average'
  )
  administration = models.TextField(max_length=2048)
  administration_rating = default_char_field(
    choices=EVAL_CHOICES,
    default='average',
    voidable=True
  )
  professionalism = models.TextField(max_length=2048)
  professionalism_rating = default_char_field(
    choices=EVAL_CHOICES,
    default='average',
    voidable=True
  )
  marketing = models.TextField(max_length=2048, **VOID)
  marketing_rating = default_char_field(
    choices=EVAL_CHOICES,
    default='average',
    voidable=True
  )
  overall_rating = default_char_field(
      choices=EVAL_CHOICES,
      default='average',
      voidable=True
  )
  rating = models.JSONField(**VOID)
  created_by = default_user_creation("%(class)s_created_by")

File: _inc/laravel/app/Models/Abstracts/Entities/Rateable.php
<?php

namespace App\Models;

use App\Models\User;
use Illuminate\Database\Eloquent\{Factories\HasFactory, Model, Relations\BelongsTo};
use Illuminate\Support\Str;

abstract class Rateable extends Model
{
  use HasFactory;
  public bool $incrementing = false;
  protected string $keyType = 'string';
  protected array $fillable = [
    'administration',
    'administration_rating',
    'attendance',
    'attendance_rating',
    'customer_experience',
    'customer_experience_rating',
    'integrity',
    'integrity_rating',
    'marketing',
    'marketing_rating',
    'overall_rating',
    'rateable_created_by',
    'rating',
  ];

  public function creator(): BelongsTo
  {
    return $this->belongsTo(User::class, 'rateable_created_by');
  }

  protected static function booted(): void
  {
    parent::booted();
    static::creating(function (self $model): void {
      $model->{$model->getKeyName()} = (string) Str::uuid();
    });
  }
}
